### ISSUE TITLE
Support `Out` keyword

### ISSUE THREAD

**Issue by bclothier** [WaynePhillipsEA/twinbasic#34]
**Describe the solution you'd like**
A useful syntactical sugar to have would be to mark a parameter as `Out`. twinBASIC allows a parameter to be passed `ByVal` or `ByRef`. However, I've often found myself coding this pattern in VBA:
```
Dim MyThing As Something
If TryGetSomething(OutSomething:=MyThing) Then
  'It's safe to access the MyThing variable here.
Else
  'I can't get to MyThing so I cannot use it. Do something without it.
End If
```

The advantage of coding in this matter is that the `TryGet*` neatly abstracts out the mechanics of getting the thing which could potentially fail (may not exist, is not available, whatever). However, there is no compile-time validation that the variable is correctly captured and corresponds to the intention of the code. This would be a misuse:

```
Dim MyThing As Something
MyThing As New Something
If TryGetSomething(MyThing) Then
  'Did MyThing mutate?
Else
  'I still have a valid MyThing instance?
End If
```

The `Out` parameter would disallow above and make it a compile-time error. 

We would then declare it as:
```Public Function TryGetSomething(Out TheThing As Something) As Boolean```

and we can then use it in 3 ways:

Use unassigned local variable
```
Dim MyThing As Something
If TryGetSomething(Out MyThing)...```

Or declare it inline:
```If TryGetSomething(Out Dim MyThing As Something)...```

Or if we don't care about the out variable:
```If TryGetSomething(Out _)...```

This would help the API become easier to understand and discoverable, even at the call sites so it's obvious when reading the code, we can see that it's expected to get something out as a parameter.

**Comment by mansellan** [comment]
Along similar lines, C# (from 7.2) allows `in` params (passed by reference, but not allowed to mutate in the called function).

**Comment by A9G-Data-Droid** [comment]
It sounds like you are just wanting to avoid a null check. Wouldn't a null coalescing operator be a better solution?

```
Dim MyThing As Something
Set MyThing = GetSomething ?? SomethingElseEntirely
```

OR

```
Dim MyThing As Something
Set MyThing = GetSomething 
Dim DependantThing As SomethingElse
Set DependantThing = MyThing?.Property
```

**Comment by bclothier** [comment]
While this is a great feature and should be its own feature request, no, the `Out` wouldn't be used for that. Not all retrieval are safe and may cause an error.

Take `DAO.Property` for example. We can use custom properties but they may or may not exist and that requires error handling to find out.

Ugly way:
```
Dim prp As DAO.Property
On Error Resume Next
Set prp = CurrentDb.Properties("SomeCustomProperty")
If Err.Number Then
  Err.Clear
  Set prp = CurrentDb.Properties.Append CurrentDb.CreateProperty("SomeCustomProperty", dbText, vbNullString)
End If
On Error GoTo 0
prp.Value = "Hi, there"
```

Better way:
```
Dim prp As DAO.Property
Dim prpName As String
Dim prpValue As String

prpName = "SomeCustomProperty"
prpValue = "hi, there"

If TryGetProperty(CurrentDb.Properties, prpName, prp) Then
  prp.Value = prpValue
Else
  CurrentDb.Properties.Append prpName, dbText, prpValue
End If
```

Ideal:
```
If TryGetProperty(CurrentDb.Properties, prpName, Out prp) Then
```

**Comment by A9G-Data-Droid** [comment]
So couldn't you do:

```
Dim prp As DAO.Property
Dim prpName As String
Dim prpValue As String

prpName = "SomeCustomProperty"
prpValue = "hi, there"

Set prp = GetProperty(CurrentDb.Properties, prpName) ?? NewProperty(CurrentDb.Properties, prpName, dbText)
prp.Value = prpValue
```

Or even:

```
With GetProperty(CurrentDb.Properties, prpName) ?? NewProperty(CurrentDb.Properties, prpName, dbText)
    .Value = prpValue
End With
```

**Comment by EduardoVB** [comment]
Hello. As for the DAO example, IMO the problem is that CurrentDb.Properties does not have an **Exists** method (many MS collections lack of it). That's would be the right way to handle that situation IMO.

About the "TryGetSomething(Out...". Can't that be handled with Is Nothing?

```
Dim MyThing As Something
If Not MyThing Is Nothing Then
  'It's safe to access the MyThing variable here.
Else
  'I can't get to MyThing so I cannot use it. Do something without it.
End If
```
How a parameter declares as "Out" would differ from a one ByRef?

**Comment by bclothier** [comment]
I agree that it'd be the right thing but alas, we can't get them to add it and we may not have the option of adding it unless tB implements something akin to extension methods (which is another scenario entirely).

In C#, the `out` enforces that the variable must be left uninitialized and also that the function assigns to it. There are no such checks with `ref` (a la `ByRef`).

Therefore this code will not compile:

```
var foo = 123;
GetARandomNumber(out foo);
```

Nor will this compile:
```
public bool GetANumber(out int someNumber)
{
     return true;
}
```

It's those compile-time checks that I want as that helps with guaranteeing the program's correctness.

**Comment by EduardoVB** [comment]
I get the point. It is more informative (seeing the Out keyword in the code) and error checking (raising an error if it was assigned beforehand or not assigned by the function) than allowing something new.

**Comment by wqweto** [comment]
In VB6 `ByVal` parameters produce `[in]` attribute in the typelib while `ByRef` produce `[in, out]`.

IMO an additional `Out` keyword would be useful to be able to produce `[out]` only parameters too.

The point being that currently VB6 can consume methods with `[out]` parameters ok and the compiler knows how to call these -- it just makes sure that the actual argument is cleared before passing it i.e. currently in VB6 using this typelib
```
    [odl, uuid(db6f6ddb-ac77-4e88-8253-819df9bbf140)]
    interface ID3D11Device : IUnknown {
        ...
        void GetImmediateContext(
            [out] ID3D11DeviceContext **ppImmediateContext);
        ...
    };
```
. . . in this VB6 code
```
    Dim pDevice As ID3D11Device
    Dim pContext As ID3D11DeviceContext
    
    Set pContext = GetNonNullContext()
    pDevice.GetImmediateContext pContext
```
. . . before calling `GetImmediateContext` the compiler emits extra `pContext` set to Nothing at callsite to explicitly clear the argument for the `[out]` only parameter so even if `GetImmediateContext` does not set/touch its `[out]` only parameter the value is never "undefined" after the actual method call.

This is much simpler (and sane) philosophy than .Net stance on the subject, keeping in mind that VB6 automatically clears local variables too so we are not used to having "undefined" state in variables/parameters at all and this has worked pretty well to simplify the language for newcommers.

**Comment by WaynePhillipsEA** [comment]
@wqweto Just FYI, that is exactly how we handle [out] args in twinBASIC too, and I agree that is the most sane way of doing it.

I also agree that having an Out keyword in tB would be a useful tool.

**Comment by Kr00l** [comment]
I suggest the keyword 'ByOut'.
This sounds just more "VB-Ish" and to honor the 'ByVal' and 'ByRef' keywords.

**Comment by EduardoVB** [comment]
The naming of things in VB has a meaning and a sense in English. 
It should be Out or ForOut IMO.

**Comment by retailcoder** [comment]
VB.NET doesn't have an equivalent for the C# `out` keyword, as per [this answer on Stack Overflow](https://stackoverflow.com/q/4358742/1188513). This part is particularly interesting:

> If you examine code in the framework (for example `Double.TryParse`), you may see the `<OutAttribute>` added to parameters, but that only makes a difference when the call is marshalled for COM interop or platform invoke.

I'd much rather have an `Out` modifier (doesn't need any more characters than that) than to introduce a whole attribute mechanism just for that.

I wish it just looked like this:

```vb
Public Function TryParse(ByVal Value As String, Out Result As Double) As Boolean
End Function
```

And then twinBASIC would simply provide compile-time assurance that `Result` is assigned a value in every execution path in the implementation of this `TryParse` method.

**Comment by retailcoder** [comment]
Actually there *is* an existing attribute syntax/mechanism... I'm just not sure it works for parameters, but... extending it could be another possibility:

```vb
Attribute TryParse.Result.TB_OutParam
```

I find an `Out` keyword more user-friendly, though.

**Comment by EduardoVB** [comment]
Another consideration is not to add many keywords that cannot later be used in code to name variables because they become reserved words.

**Comment by EduardoVB** [comment]
Some alternatives to consider:
RefOut
ByRefOut
OutRef

**Comment by retailcoder** [comment]
> And BTW, the addition of these new keywords can break existent code and backward compatibility.

Not if this parser works similar to Rubberduck's. An `Out` keyword would parse as some `OUTMOD` token *in the context of* some `ParamDefinition` parser rule, and encountered anywhere else it probably matches some `SimpleNameExpression` parser rule.

Introducing a new keyword should have no impact on what identifiers are legal or not, in pretty much every context I can think off the top of my head.

**Comment by EduardoVB** [comment]
OK, good to know that the issue of keywords can work even better than VB6.
